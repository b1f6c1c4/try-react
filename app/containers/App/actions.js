import {
  TOGGLE_DRAWER,
  LOGIN_REQUEST,
  LOGIN_SUCCESS,
  LOGIN_ERROR,
} from './constants';

export function toggleDrawer() {
  return {
    type: TOGGLE_DRAWER,
  };
}

export function loginRequest() {
  return {
    type: LOGIN_REQUEST,
  };
}

export function loginSuccess(result) {
  return {
    type: LOGIN_SUCCESS,
    result,
  };
}

export function loginError(error) {
  return {
    type: LOGIN_ERROR,
    error,
  };
}
