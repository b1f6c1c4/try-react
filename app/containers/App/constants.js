/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */
export const DEFAULT_LOCALE = 'zh'; // When can't find browser locale
export const ROOT_LOCALE = 'en'; // Fallback when there is no translation

export const TOGGLE_DRAWER = 'App/TOGGLE_DRAWER';

export const LOGIN_REQUEST = 'App/LOGIN_REQUEST';
export const LOGIN_SUCCESS = 'App/LOGIN_SUCCESS';
export const LOGIN_ERROR = 'App/LOGIN_ERROR';
