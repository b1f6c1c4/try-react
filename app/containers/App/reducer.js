import { fromJS } from 'immutable';

import {
  TOGGLE_DRAWER,
} from './constants';

// The initial state of the App
const initialState = fromJS({
  drawerOpen: false,
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case TOGGLE_DRAWER:
      if (state.get('drawerOpen')) {
        return state
          .set('drawerOpen', false);
      }
      return state
        .set('drawerOpen', true);
    default:
      return state;
  }
}

export default appReducer;
