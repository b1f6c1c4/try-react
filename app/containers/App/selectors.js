import { createSelector } from 'reselect';

export const selectGlobal = /* istanbul ignore next */ (state) => state.get('global');

export const selectDrawerOpen = createSelector(
  selectGlobal,
  /* istanbul ignore next */
  (state) => state.get('drawerOpen'),
);

export const selectJWT = createSelector(
  selectGlobal,
  /* istanbul ignore next */
  (state) => state.get('JWT'),
);
