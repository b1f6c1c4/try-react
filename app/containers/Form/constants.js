// Actions
export const DEFAULT_ACTION = 'Form/DEFAULT_ACTION';

// Sagas
export const EXTERNAL_REQUEST = 'Form/EXTERNAL_REQUEST';
export const EXTERNAL_SUCCESS = 'Form/EXTERNAL_SUCCESS';
export const EXTERNAL_FAILURE = 'Form/EXTERNAL_FAILURE';
