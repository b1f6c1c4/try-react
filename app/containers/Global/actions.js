import {
  DEFAULT_ACTION,
  EXTERNAL_REQUEST,
  EXTERNAL_SUCCESS,
  EXTERNAL_FAILURE,
} from './constants';

export function defaultAction() {
  return {
    type: DEFAULT_ACTION,
  };
}


export function externalRequest() {
  return {
    type: EXTERNAL_REQUEST,
  };
}

export function externalSuccess(result) {
  return {
    type: EXTERNAL_SUCCESS,
    result,
  };
}

export function externalFailure(error) {
  return {
    type: EXTERNAL_FAILURE,
    error,
  };
}
