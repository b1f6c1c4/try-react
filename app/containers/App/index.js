/**
 *
 * App.js
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 *
 * NOTE: while this component should technically be a stateless functional
 * component (SFC), hot reloading does not currently support SFCs. If hot
 * reloading is not a necessity for you then you can refactor it and remove
 * the linting exception.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createSelector, createStructuredSelector } from 'reselect';
import { Switch, Route } from 'react-router-dom';
import { withStyles, Drawer, List, Hidden, Divider, AppBar, Toolbar, Typography, Button, IconButton } from 'material-ui';
import MenuIcon from 'material-ui-icons/Menu';

import HomePage from 'containers/HomePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import { toggleDrawer } from './actions';

const drawerWidth = 240;

const styles = (theme) => ({
  root: {
    width: '100%',
    zIndex: 1,
    overflow: 'hidden',
  },
  appFrame: {
    position: 'relative',
    display: 'flex',
    width: '100%',
    height: '100%',
  },
  appBar: {
    position: 'absolute',
    marginLeft: drawerWidth,
    [theme.breakpoints.up('md')]: {
      width: `calc(100% - ${drawerWidth}px)`,
    },
  },
  navIconHide: {
    [theme.breakpoints.up('md')]: {
      display: 'none',
    },
  },
  drawerHeader: theme.mixins.toolbar,
  drawerPaper: {
    width: 250,
    [theme.breakpoints.up('md')]: {
      width: drawerWidth,
      position: 'relative',
      height: '100%',
    },
  },
  content: {
    backgroundColor: theme.palette.background.default,
    width: '100%',
    padding: theme.spacing.unit * 3,
    height: 'calc(100% - 56px)',
    marginTop: 56,
    [theme.breakpoints.up('sm')]: {
      height: 'calc(100% - 64px)',
      marginTop: 64,
    },
  },
  flex: {
    flex: 1,
  },
});

class App extends React.PureComponent {
  render() {
    const { classes, theme } = this.props;

    const drawer = (
      <div>
        <div className={classes.drawerHeader} />
        <Divider />
        <List></List>
      </div>
    );

    return (
      <div className={classes.root}>
        <div className={classes.appFrame}>
          <AppBar className={classes.appBar}>
            <Toolbar>
              <IconButton
                color="contrast"
                aria-label="menu"
                onClick={this.props.onDrawerToggle}
                className={classes.navIconHide}
              >
                <MenuIcon />
              </IconButton>
              <Typography className={classes.flex} type="title" color="inherit" noWrap>
                Try React
              </Typography>
              <Button color="contrast">Login</Button>
            </Toolbar>
          </AppBar>
          <Hidden mdUp>
            <Drawer
              type="temporary"
              anchor={theme.direction === 'rtl' ? 'right' : 'left'}
              open={this.props.drawerOpen}
              classes={{
                paper: classes.drawerPaper,
              }}
              onRequestClose={this.props.onDrawerToggle}
              ModalProps={{
                keepMounted: true, // Better open performance on mobile.
              }}
            >
              {drawer}
            </Drawer>
          </Hidden>
          <Hidden mdDown implementation="css">
            <Drawer
              type="permanent"
              open
              classes={{
                paper: classes.drawerPaper,
              }}
            >
              {drawer}
            </Drawer>
          </Hidden>
          <main className={classes.content}>
            <Switch>
              <Route exact path="/" component={HomePage} />
              <Route component={NotFoundPage} />
            </Switch>
          </main>
        </div>
      </div>
    );
  }
}

App.propTypes = {
  classes: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired,
  drawerOpen: PropTypes.bool.isRequired,
  onDrawerToggle: PropTypes.func.isRequired,
};

export function mapDispatchToProps(dispatch) {
  return {
    onDrawerToggle: () => dispatch(toggleDrawer()),
  };
}

const mapStateToProps = createStructuredSelector({
  drawerOpen: createSelector(
    (state) => state.get('global'),
    (state) => state.get('drawerOpen'),
  ),
});

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(
  withConnect,
  withStyles(styles, { withTheme: true }),
)(App);
