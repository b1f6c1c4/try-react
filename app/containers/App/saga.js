import { call, put, select, takeLatest } from 'redux-saga/effects';
import { LOGIN_REQUEST } from 'containers/App/constants';
import { loginSuccess, loginError } from 'containers/App/actions';

import { apiPOST } from 'utils/request';

export function* loginRequest() {
  const json = yield select((store) => store.get('global').get('credential'));

  try {
    const result = yield call(apiPOST, '/login', undefined, json);
    yield put(loginSuccess(result));
  } catch (err) {
    yield put(loginError(err));
  }
}

export default function* login() {
  yield takeLatest(LOGIN_REQUEST, loginRequest);
}
