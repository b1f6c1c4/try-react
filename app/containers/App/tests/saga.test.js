import { put, takeLatest } from 'redux-saga/effects';

import { LOGIN_REQUEST } from 'containers/App/constants';
import { loginSuccess, loginError } from 'containers/App/actions';

import login, { loginRequest } from '../saga';

const credential = 'the-credential';

/* eslint-disable redux-saga/yield-effects */
describe('loginRequest Saga', () => {
  let loginRequestGenerator;

  beforeEach(() => {
    loginRequestGenerator = loginRequest();

    const selectDescriptor = loginRequestGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = loginRequestGenerator.next(credential).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the loginSuccess action if it requests the data successfully', () => {
    const response = [{
      name: 'First',
    }, {
      name: 'Second',
    }];
    const putDescriptor = loginRequestGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(loginSuccess(response)));
  });

  it('should call the loginError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = loginRequestGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(loginError(response)));
  });
});

describe('login Saga', () => {
  const saga = login();

  it('should start task to watch for LOGIN_REQUEST action', () => {
    const takeLatestDescriptor = saga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOGIN_REQUEST, loginRequest));
  });
});
