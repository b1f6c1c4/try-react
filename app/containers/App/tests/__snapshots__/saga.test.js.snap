// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginRequest Saga should call the loginError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`loginRequest Saga should call the loginError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/login",
      undefined,
      "the-credential",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`loginRequest Saga should dispatch the loginSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`loginRequest Saga should dispatch the loginSuccess action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/login",
      undefined,
      "the-credential",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
